### PR TITLE
NWB ingestion

### LINKED ISSUE #37: Volume dimensions order
The current code seems to presume a specific order of dimensions `(z, y, x)` for the input volumes:

https://github.com/mikarubi/voluseg/blob/3b8465d710815eac7733d681880a67eef891bc73/voluseg/_steps/step1.py#L59

But data coming from other sources might have a different order. For example, in NWB the order of dimensions should be `(t, x, y)` or `(t, x, y, z)`, [see here](https://pynwb.readthedocs.io/en/stable/pynwb.ophys.html#pynwb.ophys.TwoPhotonSeries).

One idea is to have a parameters string argument `dimensions_order` which would:
- have a default value of `z,y,x`, as it is presumed today
- enforce the NWB order `x,y,z` if `ext==".nwb"`

any other suggestion, @mikarubi ?

### PR DESCRIPTION
**Pull request by luiztauffer**
Ingestion of [TwoPhotonSeries](https://pynwb.readthedocs.io/en/stable/tutorials/domain/ophys.html#two-photon-series) volumes from local NWB files.

- [x] modifications on step0 and step1 to ingest NWB data
- [x] auxiliary functions to handle NWB files
- [x] read remote NWB files from a S3 url
- [x] auxiliary function to download sample dataset in the NWB format [from here](https://gui-staging.dandiarchive.org/dandiset/215495?pos=1)
- [x] Fix #37 
- [x] extended tests to run pipeline for NWB and compare with results from h5 directory format

### CODE REVIEW

**Comment by luiztauffer** [comment]
@mikarubi there are at least two points we should discuss for this first part. It would be great if you could clarify to me how you see these points being solved:

**1. Parameters dictionary**
To allow voluseg to keep reading from a list of files, stored in a list of directories (current behavior), and at the same time allow for it to ingest all the data from a single NWB file (what is typically the case), we might need to extend the parameters.
The simplest thing would be to include:
- `nwb_input_local_path`: the path to a local NWB file. Finding it automatically from a folder is of course possible, but it is common that the same folder contains multiple NWB files (dandisets are organized like this for example), and in that case the `dir_input` would be ambiguous.
- `nwb_input_acquisition_name`: the name of the TwoPhotonSeries object inside the NWB file acquisition path, which might contain multiple acquisitions. This we can certainly infer automatically, and raise an error if the number of  TwoPhotonSeries in the file is not == 1. This can be one of the processed parameters, so the user doesn’t need to provide it, but it will be useful to have in the parameters dictionary for the next steps in the voluseg pipeline.

**2. Ingestion of NWB data**
Besides the different reading methods, the volumes inside the `.nwb` files could be treated exactly the same way as the volumes stored in the `.h5` files. This includes e.g. downsampling from the step 1. My idea is that we could provide an initial interface for reading from different input files (`.h5`, `.nwb`) but that subsequent steps (1-5) should be agnostic to the original file format, that helps a lot with modularizing, growing and maintaining the project. 
But by your comments on the first PR, I understood that you wanted data coming from NWB to take a different processing path, or did I misunderstand you?

**Comment by mikarubi** [comment]
> @mikarubi there are at least two points we should discuss for this first part. It would be great if you could clarify to me how you see these points being solved:
> 
> **1. Parameters dictionary** To allow voluseg to keep reading from a list of files, stored in a list of directories (current behavior), and at the same time allow for it to ingest all the data from a single NWB file (what is typically the case), we might need to extend the parameters. The simplest thing would be to include:

Here, I think we should simply extend the `dir_input` functionality as follows:

`dir_input = "/path/to/external/nwb_file.nwb:/path/to/internal/dir"`

i.e. the user separates the external and internal directories with a semicolon which we can easily parse. This is consistent with the role of `input_dir` (specifying directory to location of data) but just extends it to the hdf5 realm.

> **2. Ingestion of NWB data** Besides the different reading methods, the volumes inside the `.nwb` files could be treated exactly the same way as the volumes stored in the `.h5` files. This includes e.g. downsampling from the step 1. My idea is that we could provide an initial interface for reading from different input files (`.h5`, `.nwb`) but that subsequent steps (1-5) should be agnostic to the original file format, that helps a lot with modularizing, growing and maintaining the project. But by your comments on the first PR, I understood that you wanted data coming from NWB to take a different processing path, or did I misunderstand you?

Yes agree, that's probably the simplest approach that preserves the full functionality (including registration). So let's go with that.

**Comment by luiztauffer** [comment]
@mikarubi this is good for review

File: voluseg/_steps/step1.py [hunk 1]
```diff
@@ -56,7 +61,14 @@ def make_output_volume(name_volume, volume):
                 # fix dimensionality
                 if volume.ndim == 2:
                     volume = volume[None, :, :]
-                volume = volume.transpose(2, 1, 0)
+
+                # reorder dimensions to xyz, if necessary
+                if p.dim_order == "zyx":
```
Comment (mikarubi) [inline voluseg/_steps/step1.py:66]: Minor comment. At some stage we should add all possible orders here:
* xyz
* xzy
* yxz
* yzx
* zxy
* zyx

File: voluseg/_steps/step0.py [hunk 2]
```diff
@@ -144,48 +145,75 @@ def process_parameters(initial_parameters: dict) -> dict:
...
-        volume_names_input_h = sorted([i for i, j in file_names if "." + j == ext])
-        volume_fullnames_input_h = [
-            os.path.join(dir_input_h, i) for i in volume_names_input_h
-        ]
+    if (".nwb" in dir_input) or ("https://" in dir_input):
+        if "https://" in dir_input:
+            remote = True
+            aux_list = [dir_input]
+        else:
+            aux_list = dir_input.split(":")
```
Comment (mikarubi) [inline voluseg/_steps/step0.py:156]: In principle, we could use a similar approach for other 4D datasets (e.g. zarr).

**Review by mikarubi — APPROVED** [review]
Looks good in general, I left a couple of minor comments (not for this PR, but perhaps for the future).

**Pull request merged**